File: Leetcode 2601-2700/2685/2685.java
// Leetcode 2685: Count the Number of Complete Components
// https://leetcode.com/problems/count-the-number-of-complete-components/
// Solved on 11th of July, 2026
import java.util.ArrayList;
import java.util.List;

class Solution {
    /**
     * Counts the number of complete components in a graph.
     * 
     * @param n The number of nodes in the graph.
     * @param edges The edges in the graph.
     * @return The number of complete components.
     */
    public int countCompleteComponents(int n, int[][] edges) {
        List<Integer>[] adjList = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adjList[i] = new ArrayList<>();
        }
        for (int[] edge : edges) {
            adjList[edge[0]].add(edge[1]);
            adjList[edge[1]].add(edge[0]);
        }
        boolean[] visited = new boolean[n];
        int completeComponents = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                int[] stats = new int[2];
                dfs(i, adjList, visited, stats);
                if (stats[1] == stats[0] * (stats[0] - 1)) {
                    completeComponents++;
                }
            }
        }
        return completeComponents;
    }

    private void dfs(int node, List<Integer>[] adjList, boolean[] visited, int[] stats) {
        visited[node] = true;
        stats[0]++;
        stats[1] += adjList[node].size();
        for (int neighbor : adjList[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor, adjList, visited, stats);
            }
        }
    }
}

File: Leetcode 2601-2700/2685/2685.py
# Leetcode 2685: Count the Number of Complete Components
# https://leetcode.com/problems/count-the-number-of-complete-components/
# Solved on 11th of July, 2026
class Solution:
    def countCompleteComponents(self, n: int, edges: list[list[int]]) -> int:
        """
        Counts the number of complete components in a graph.

        Args:
            n: The number of nodes in the graph.
            edges: The edges in the graph.

        Returns:
            The number of complete components.
        """
        adjList = [[] for _ in range(n)]
        for u, v in edges:
            adjList[u].append(v)
            adjList[v].append(u)
        
        visitedNodes = [False] * n
        completeComponentCount = 0

        for i in range(n):
            if not visitedNodes[i]:
                componentNodes = []
                nodeStack = []
                visitedNodes[i] = True

                while nodeStack:
                    currentNode = nodeStack.pop()
                    componentNodes.append(currentNode)
                    for neighborNode in adjList[currentNode]:
                        if not visitedNodes[neighborNode]:
                            visitedNodes[neighborNode] = True
                            nodeStack.append(neighborNode)
                
                nodeCount = len(componentNodes)
                isComplete = True
                for node in componentNodes:
                    if len(adjList[node]) != nodeCount - 1:
                        isComplete = False
                        break
                
                if isComplete:
                    completeComponentCount += 1
        
        return completeComponentCount